Mathlib/LinearAlgebra/LinearDisjoint.lean

```diff
@@ -0,0 +1,550 @@
+  are commutative, then the rank of the intersection of `M` and `N` is at most one.
+
+- `Submodule.LinearDisjoint.rank_le_one_of_commute_of_flat_of_self`:
+  if `M` and itself are linearly disjoint, if `M` is flat, if any two elements in `M`
+  are commutative, then the rank of `M` is at most one.
+
+The results with name containing "of_commute" also have corresponding specified versions
+assuming `S` is commutative.
+
+## Tags
+
+linearly disjoint, linearly independent, tensor product
+
+-/
+
+open scoped Classical TensorProduct
+
+noncomputable section
+
+universe u v w
+
+namespace Submodule
+
+variable {R : Type u} {S : Type v}
+
+section Semiring
+
+variable [CommSemiring R] [Semiring S] [Algebra R S]
+
+variable (M N : Submodule R S)
+
+/-- Two submodules `M` and `N` in an algebra `S` over `R` are linearly disjoint if the natural map
+`M ⊗[R] N →ₗ[R] S` induced by multiplication in `S` is injective. -/
+@[mk_iff]
+protected structure LinearDisjoint : Prop where
+  injective : Function.Injective (mulMap M N)
+
+variable {M N}
+
+/-- If `M` and `N` are linearly disjoint submodules, then there is the natural isomorphism
+`M ⊗[R] N ≃ₗ[R] M * N` induced by multiplication in `S`. -/
+protected def LinearDisjoint.mulMap (H : M.LinearDisjoint N) : M ⊗[R] N ≃ₗ[R] ↥(M * N) :=
+  LinearEquiv.ofInjective (M.mulMap N) H.injective ≪≫ₗ LinearEquiv.ofEq _ _ (mulMap_range M N)
+
+@[simp]
+theorem LinearDisjoint.val_mulMap_tmul (H : M.LinearDisjoint N) (m : M) (n : N) :
+    (H.mulMap (m ⊗ₜ[R] n) : S) = m.1 * n.1 := rfl
+
+@[nontriviality]
+theorem LinearDisjoint.of_subsingleton [Subsingleton R] : M.LinearDisjoint N := by
+  haveI : Subsingleton S := Module.subsingleton R S
+  exact ⟨Function.injective_of_subsingleton _⟩
+
+/-- Linearly disjoint is preserved by taking multiplicative opposite. -/
+theorem linearDisjoint_op :
+    M.LinearDisjoint N ↔ (equivOpposite.symm (MulOpposite.op N)).LinearDisjoint
+      (equivOpposite.symm (MulOpposite.op M)) := by
+  simp only [linearDisjoint_iff, mulMap_op, LinearMap.coe_comp,
+    LinearEquiv.coe_coe, EquivLike.comp_injective, EquivLike.injective_comp]
+
+alias ⟨LinearDisjoint.op, LinearDisjoint.of_op⟩ := linearDisjoint_op
+
+/-- Linearly disjoint is symmetric if elements in the module commute. -/
+theorem LinearDisjoint.symm_of_commute (H : M.LinearDisjoint N)
+    (hc : ∀ (m : M) (n : N), Commute m.1 n.1) : N.LinearDisjoint M := by
+  rw [linearDisjoint_iff, mulMap_comm_of_commute M N hc]
+  exact ((TensorProduct.comm R N M).toEquiv.injective_comp _).2 H.injective
+
+/-- Linearly disjoint is symmetric if elements in the module commute. -/
+theorem linearDisjoint_symm_of_commute
+    (hc : ∀ (m : M) (n : N), Commute m.1 n.1) : M.LinearDisjoint N ↔ N.LinearDisjoint M :=
+  ⟨fun H ↦ H.symm_of_commute hc, fun H ↦ H.symm_of_commute fun _ _ ↦ (hc _ _).symm⟩
+
+namespace LinearDisjoint
+
+variable (M N)
+
+theorem of_map_linearIndependent_left' {ι : Type*} (m : Basis ι R M)
```

Comment:
I think the name should mention `Basis` instead of `linearIndependent`.

Reply:
Done.